### PR TITLE
cli: misc cleanups

### PR DESCRIPTION
**Pull request by andreimatei**
See the individual commits.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10303)

### CODE REVIEW

**Comment by andreimatei** [comment]
I need to figure out how to comply exactly with the BSD license of the code from Go's sql package that I've copied.

**Comment by petermattis** [comment]
> I need to figure out how to comply exactly with the BSD license of the code from Go's sql package that I've copied.

I think you need to maintain the copyright comment at the top of the file. See the comments at the top of `sql/parser/create.go` for an example.

**Comment by andreimatei** [comment]
I've copied the copyright and the license from the Go source file. But I think we also need to do something for "distribution in binary form"...?
`sql/parser/create.go` says that the BSD license is in the `LICENSE` file, but I don't see it...

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10303#-:-KVQxyt5FaWIdnQEIy_-:b1dz386)*

File: pkg/cli/dump.go [hunk 1]
```diff
@@ -88,10 +88,12 @@ func dumpTable(w io.Writer, conn *sqlConn, origDBName, origTableName string) err
 	if err != nil {
 		return err
 	}
-	vals = make([]driver.Value, 2)
+	defer func() { _ = rows.Close() }()
```
Comment (mberhault) [inline pkg/cli/dump.go]: you're calling `Close` twice on this instance of `rows`, in the defer and at L113.
While it's probably no big deal, it feels like this part of the function could be moved to a helper function.

File: pkg/cli/dump.go [hunk 2]
```diff
@@ -119,11 +121,13 @@ func dumpTable(w io.Writer, conn *sqlConn, origDBName, origTableName string) err
 	if err != nil {
 		return err
 	}
-	vals = make([]driver.Value, 5)
+	defer func() { _ = rows.Close() }()
 	var primaryIndex string
 	// Find the primary index columns.
-	for {
-		if err := rows.Next(vals); err == io.EOF {
+	for rows.Next() {
```
Comment (mberhault) [inline pkg/cli/dump.go]: ditto here and L150

File: pkg/cli/dump.go [hunk 3]
```diff
@@ -198,8 +202,9 @@ func dumpTable(w io.Writer, conn *sqlConn, origDBName, origTableName string) err
 		inserts := make([][]string, 0, limit)
 		i := 0
 		for i < limit {
-			vals := make([]driver.Value, len(cols)+len(pkcols))
-			if err := rows.Next(vals); err == io.EOF {
+			rows.Next()
```
Comment (mberhault) [inline pkg/cli/dump.go]: this instance of `rows` (created at L195) doesn't get a defer. And we definitely have return statements in the middle.

**Comment by tamird** [comment]
Before we go vendoring chunks of the standard library: have you considered using https://github.com/jackc/pgx for those places where it matters?

**Comment by andreimatei** [comment]
I looked at pgx a bit. It doesn't seem to support multiple result sets, which the CLI needs. I supposed everybody's gonna get that support soon though, as Go 1.8 will provide a standard interface for it. At that point maybe we won't need to maintain all the crap that we already have around libpq, so maybe that's gonna be a good time for tests that want pinning to a connection to use something else.
In the meantime, I'm not "vendoring" all that much - just one rando function.

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/cli/dump.go, line 91 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10303#-KVCE8FI4prsn3YK7qMO-r1-91:-KVk1vy7W7lvdOZt8jD_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd913f124480231f2a62617306fd0cac3498bca/pkg/cli/dump.go#L91)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> you're calling `Close` twice on this instance of `rows`, in the defer and at L113.
> 
> While it's probably no big deal, it feels like this part of the function could be moved to a helper function.
</details>
Done.

---

*[pkg/cli/dump.go, line 127 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10303#-KVCE8FI4prsn3YK7qMO-r1-127:-KVk1wloF_axH-Z-2VhY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd913f124480231f2a62617306fd0cac3498bca/pkg/cli/dump.go#L127)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> ditto here and L150
</details>
Done.

---

*[pkg/cli/dump.go, line 205 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10303#-KVCE8FI4prsn3YK7qMO-r1-205:-KVkRO3mZUiafSlKndNC:b2u0sat) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd913f124480231f2a62617306fd0cac3498bca/pkg/cli/dump.go#L205)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> this instance of `rows` (created at L195) doesn't get a defer. And we definitely have return statements in the middle.
</details>
Done.

I've refactored the thing a bit to make it more legible, in a second commit.
Also triggered a cockroach panic in the process :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10303#-:-KVkT9SXMJp63BjtNnj6:bp6ty63)*

**Comment by mjibson** [comment]
FYI lib/pq recently got multiple result sets, which we can use as soon as we move to a go 1.8 RC, or we do something crazy like vendor database/sql from 1.8.

**Comment by andreimatei** [comment]
I've removed the vendoring of the gosql function that I've had and was convinced to not use cli.sqlRows outside of the CLI package. What remains here are some cleanups that I think are worthwhile.

---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10303#-:-KaeLOXs5eOdwTimqtrG:b-biyv2d)*

**Comment by knz** [comment]
@andreimatei I think the code has evolved in a similar direction since. Do you think this PR is still worthwhile?

**Pull request closed without merging**